File: app/src/main/java/com/pr0gramm/app/ui/DetectTapTouchListener.java
package com.pr0gramm.app.ui;

import android.view.MotionEvent;
import android.view.View;

import rx.functions.Action0;

/**
 * Detects single taps.
 */
public class DetectTapTouchListener implements View.OnTouchListener {
    private final Action0 consumer;
    private boolean moveOccurred;
    private float firstX;
    private float firstY;

    private DetectTapTouchListener(Action0 consumer) {
        this.consumer = consumer;
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        if (event.getActionIndex() == 0) {
            if (event.getAction() == MotionEvent.ACTION_DOWN) {
                moveOccurred = false;
                firstX = event.getX();
                firstY = event.getY();
            }

            if (event.getAction() == MotionEvent.ACTION_MOVE) {
                moveOccurred = moveOccurred
                        || Math.abs(event.getX() - firstX) > 32
                        || Math.abs(event.getY() - firstY) > 32;
            }

            if (event.getAction() == MotionEvent.ACTION_UP) {
                if (!moveOccurred) {
                    consumer.call();
                    return true;
                }
            }
        }

        return false;
    }

    public static DetectTapTouchListener withConsumer(Action0 consumer) {
        return new DetectTapTouchListener(consumer);
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/DetectTapTouchListener.kt
package com.pr0gramm.app.ui

import android.view.MotionEvent
import android.view.View

/**
 * Detects single taps.
 */
class DetectTapTouchListener(private val consumer: () -> Unit) : View.OnTouchListener {
    private var moveOccurred: Boolean = false
    private var firstX: Float = 0.toFloat()
    private var firstY: Float = 0.toFloat()

    override fun onTouch(v: View, event: MotionEvent): Boolean {
        if (event.actionIndex == 0) {
            if (event.action == MotionEvent.ACTION_DOWN) {
                moveOccurred = false
                firstX = event.x
                firstY = event.y
            }

            if (event.action == MotionEvent.ACTION_MOVE) {
                moveOccurred = moveOccurred
                        || Math.abs(event.x - firstX) > 32
                        || Math.abs(event.y - firstY) > 32
            }

            if (event.action == MotionEvent.ACTION_UP) {
                if (!moveOccurred) {
                    consumer()
                    return true
                }
            }
        }

        return false
    }
}
